game-core/src/main/java/games/strategy/triplea/ui/unit/scroller/UnitScroller.java

```diff
@@ -241,39 +247,81 @@ public void centerOnCurrentMovableUnit() {
     }
   }
 
+  private List<Unit> getMovableUnits(final Territory territory) {
+    if (territory == null) {
+      return Collections.emptyList();
+    }
+    return UnitScrollerModel.getMoveableUnits(
+        territory, movePhaseSupplier.get(), currentPlayerSupplier.get(), getAllSkippedUnits());
+  }
+
   /**
    * Skips the units in the current territory. Scroller will not scroll back to these units and the
    * units will not be highlighted.
    */
   public void skipCurrentUnits() {
     if (lastFocusedTerritory != null) {
-      skippedUnits.addAll(lastFocusedTerritory.getUnits());
+      skippedUnits.addAll(getMovableUnits(lastFocusedTerritory));
       updateMovesLeftLabel();
     }
     centerOnNextMovableUnit();
   }
 
   public void sleepCurrentUnits() {
     if (lastFocusedTerritory != null) {
-      sleepingUnits.addAll(lastFocusedTerritory.getUnits());
+      sleepingUnits.addAll(getMovableUnits(lastFocusedTerritory));
       updateMovesLeftLabel();
     }
     centerOnNextMovableUnit();
   }
 
   public void centerOnNextMovableUnit() {
     centerOnMovableUnit(true);
+
+    if (lastFocusedTerritory != null) {
+      showAllUnitsMovedIfNeeded();
+    }
+  }
+
+  private void showAllUnitsMovedIfNeeded() {
+    if (movesLeft() == 0 && ClientSetting.notifyAllUnitsMoved.getValueOrThrow()) {
+      showAllUnitsMoved();
+    }
+  }
+
+  private void showAllUnitsMoved() {
+    final int result =
+        JOptionPane.showOptionDialog(
+            mapPanel,
+            "All units have moved",
+            "All units moved",
+            JOptionPane.YES_NO_OPTION,
+            JOptionPane.PLAIN_MESSAGE,
+            null,
+            new String[] {"OK", "Do not show again"},
+            "OK");
+    if (result == 1) {
```

Comment:
JOptionPane should have constants for this.
I'm pretty sure intellij has a quick fix for this, but maybe it doesn't work because the assignment and the actual comparison aren't the same. But don't know

Reply:
Swing is returning the index position of the array passed in as the second to the last argument. The array is arbitrary, so there couldn't be constants for arbitrary index positions. It is in contrast to a YES_NO dialog or the YES_NO_CANCEL dialog which IIRC does return magic constants as well. Those constants I think are also '0', '1', and '2', but the relationship there is only coincidental, they just happen to be the same value.

Reply:
Fair point, didn't know that. Perhaps this indicates it should be documented what the 1 stands for?

Reply:
It is a shortcoming of the swing API. I'm not sure how it can be commented on without more or less saying  "see the javadocs of JOptionPane"

Having a better API is probably a better answer. TBH I think this falls under the 'know your library' category. Swing has a horrible API like this.

We have other examples, should we document them all? Do you have any kind of comment in mind that would not just be copy/pasted next to every usage and be redundant to the javadoc on the API?

Examples:

ViewMenu.java:150
```
              final int result =
                  JOptionPane.showOptionDialog(
                      frame,
                      panel,
                      "Choose Map Scale",
                      JOptionPane.OK_CANCEL_OPTION,
                      JOptionPane.PLAIN_MESSAGE,
                      null,
                      new String[] {"OK", "Cancel"},
                      0);
              if (result != 0) {
                return;
              }
```

ViewMenu:382
```
              final int result =
                  JOptionPane.showOptionDialog(
                      frame,
                      ui,
                      "Edit Map Font and Color",
                      JOptionPane.YES_NO_CANCEL_OPTION,
                      JOptionPane.PLAIN_MESSAGE,
                      null,
                      options,
                      2);
              if (result == 1) {
```

Reply:
Maybe a single comment "Check if second button was pressed" seems reasonable?

If you disagree feel free to merge.

Reply:
@RoiEXLab thought of something, an intermediate variable to describe the condition seems to work, updated in: 0a98c9296